### ISSUE TITLE
Support for Nextcloud App "Fulltextsearch - Elasticsearch" https://github.com/nextcloud/fulltextsearch_elasticsearch

### ISSUE THREAD

**Issue by herrep** [nextcloud/univention-app#70]
Dear App Maintainer,

Would you consider support for integrating the Nextcloud App "Fulltextsearch - Elasticsearch" available at https://github.com/nextcloud/fulltextsearch_elasticsearch into your Nextcloud Docker Container?

When activating this Nextcloud App, I can configure it by referring to an Elasticsearch server outside the Nextcloud Docker Container, but it appears that a connection to this external reference is refused by the Nextcloud Docker Container.

If there are no plans to do so, what would be the best way forward to integrate the Nextcloud App "Fulltextsearch - Elasticsearch" or is the only way of installing Nextcloud directly within UCS 4.3?

Best regards,
Peter

**Comment by blizzz** [comment]
> Would you consider support for integrating the Nextcloud App "Fulltextsearch - Elasticsearch" available at https://github.com/nextcloud/fulltextsearch_elasticsearch into your Nextcloud Docker Container?

Would like to work in it? Then, I am open to integrate support for that.

> it appears that a connection to this external reference is refused by the Nextcloud Docker Container

I am not familiar with elastic search and the integration. This should only be a problem, if there are incoming connections towards Nextcloud on a port that is not 443 (via reverse proxy) or 80 (internally, the reverse proxy on the ucs host does ssl offloading).

**Comment by blizzz** [comment]
@herrep you might try the suggestion in https://github.com/nextcloud/fulltextsearch/issues/256#issuecomment-367584598 Essentially strip the protocol from the host.

**Comment by herrep** [comment]
Hmm... Essentially, I had not an issue with setting up Elasticsearch outside the Docker container. I bound Elasticsearch to the Public IP of the server which is also accessible from Docker so that Nextcloud inside the Docker container can reach it.

I have Elasticsearch 6.4.2 on my server and Nextcloud 13.0.6 in a Docker container.

Running a test of the fulltextsearch from inside the Nextcloud Docker container works fine:
`sudo -u www-data /var/www/html/occ fulltextsearch:test`

But when I run the index, I only see the screen below, without any progress:
`sudo -u www-data /var/www/html/occ fulltextsearch:index`

<img width="612" alt="elasticsearch_occ_index" src="https://user-images.githubusercontent.com/41866019/46617123-4449a500-cb1c-11e8-8e14-811cdd3016c8.png">

**Comment by herrep** [comment]
I cannot even leave the terminal window, neither pressing q or Ctrl-C does have an effect.

**Comment by herrep** [comment]
Elasticsearch is on my server accessible at http://MyServerLanIpAddress:9200.
From inside the Nextcloud Docker container, I get response from
`curl 'http://<MyServerLanIpAddress>:9200'`
As reported above, occ fulltextsearch:test works finem but fulltextsearch:index shows no progress.

**Comment by blizzz** [comment]
The question is, how did you configure elasticsearch within nextcloud?

**Comment by herrep** [comment]
In the Nextcloud settings, I specified the elasticsearch server with and without a http:// in front of hostname:port. This made no difference. Before this, I opened a shell in the Docker container and verified that curl can contact the elasticsearch server. Here is my Elasticsearch configuration in Nextcloud:

<img width="808" alt="elasticsearch" src="https://user-images.githubusercontent.com/41866019/46621509-ede26380-cb27-11e8-8779-39ae82e9b25c.png">

**Comment by blizzz** [comment]
@daita perhaps you've got an idea? Does it make sense to reopen the issue on https://github.com/nextcloud/fulltextsearch rather?

**Comment by ArtificialOwl** [comment]
So, the `:test` command is all green ?

How many files do you have ?
What is the setup of your files ? remote ? local ? encrypted ?

**Comment by herrep** [comment]
Yes, the test command succeeded entirely. The document server is remote to the Nextcloud server in the same LAN and hosts mainly text documents, total size around 500 GB. The files are not encrypted and accessed by ssh/sftp.

**Comment by ArtificialOwl** [comment]
500GB is nice, it can take a long time while to get the full files tree. How long have you wait before trying to kill the process ?

**Comment by herrep** [comment]
I waited long enough, as occ terminated with "Force Quit". BTW, as soon as the index screen is shown, it starts showing progress " / " without any numbers. Then, after a long time (in the range of hours) it switched to a progress of "0/274530". In the time between, the user interface commands do not appear to show a reaction.

<img width="541" alt="bildschirmfoto 2018-10-09 um 11 22 00" src="https://user-images.githubusercontent.com/41866019/46659858-0a2ce180-cbb6-11e8-8e4f-7fb95eca94b9.png">

**Comment by ArtificialOwl** [comment]
Do you have more information about the memory used by the script before crashing ? do you think you reach the limit of the assigned RAM to the PHP process ?

**Comment by herrep** [comment]
Unfortunately, I am not familiar in debugging PHP scripts. Could you point me to the right direction as how to check the RAM usage and to increase the amount of assigned memory? Then I will gladly provide the required information.

**Comment by ArtificialOwl** [comment]
It should be displayed above of the Action: line from your screenshot

**Comment by herrep** [comment]
I started the index right now, and it reports 9 MB.

**Comment by ArtificialOwl** [comment]
there is no refresh during the process of `generatingIndexFiles`, you should have another value after this step is done

**Comment by herrep** [comment]
Now occ is in the process of indexing, currently about 165,000 out of 280,000 documents processed. Memory is at 393 MB, still indexing. However, already more than 7,000 errors are thrown with the same content:

Exception: Elasticsearch\Common\Exceptions\BadRequest400Exception
Message: pipeline with id [attachment] does not exist

**Comment by ArtificialOwl** [comment]
during the index, you should be able to navigate throw the errors and get the ID of the document (files:123456)

with those information, open a new terminal and try:

./occ fultextsearch:document:platform files 123456

you should have information about the files with issues.
Are those files bigger than the limit you set up in the settings page ? 20MB by default

**Comment by ArtificialOwl** [comment]
note that this first command should give you the owner of the document, then you can get information about the file from nextcloud itself using:

./occ fulltextsearch:document:provider owner files 123456

**Comment by herrep** [comment]
I processed the first command to identify the owner:

> sudo -u www-data /var/www/html/occ fulltextsearch:document:platform files 342447
> {
>    "document": {
>        "id": "342447",
>        "providerId": "files",
>        "access": {
>            "ownerId": "",
>            "viewerId": null,
>            "users": [],
>            "groups": [],
>            "circles": [],
>            "links": []
>        },
>        "modifiedTime": 0,
>        "title": "\<filename\>",
>        "link": "",
>        "index": null,
>        "source": "files_local",
>        "info": [],
>        "hash": "\<hash\>",
>        "contentSize": 0,
>        "tags": [],
>        "metatags": [
>          "files_local"
>        ],
>        "subtags": [],
>        "more": null,
>        "excerpts": [],
>        "score": null
>    }
> }

However, I do not find any owner?!

**Comment by herrep** [comment]
(The title and the hash fields were originally filled with data which I removed for privacy reasons. The remaining output is the original output.)

**Comment by herrep** [comment]
The given error case above concerns a PDF document with 49 kByte of size and having blanks in the file name.

**Comment by herrep** [comment]
The index process ended with "Force quit" at memory usage of 519 MB
<img width="572" alt="bildschirmfoto 2018-10-18 um 11 20 07" src="https://user-images.githubusercontent.com/41866019/47144522-ec533100-d2c7-11e8-8c66-a403defb5858.png">

**Comment by ArtificialOwl** [comment]
thanks, i'll have a look to this pipeline issue, in the meanwhile, do you know if you have an idea if you have a memory limit set for your php processes ?

**Comment by blizzz** [comment]
That should be Ubuntu 16.04 default, which is 128MB on apache, unlimited on CLI

**Comment by ArtificialOwl** [comment]
The `Force Quit` thing seems to be triggered if elasticsearch takes more than 5 minutes to process the content of a file.

There is 3 reasons for this timeout: 

- You have set up a maximum file size in the Settings page which is too high (> 70MB)
- The server itself is (really) loaded.
- You killed the process using the `./occ fulltextsearch:stop` in an other terminal.

You can set the TTL (Time To Live) in the file `lib/Model/Runner.php` (L41) from the app `fulltextsearch`:
>       const TICK_TTL = 300;

This value will be set to 1800 (30 minutes) in a future release of the app.

**Comment by herrep** [comment]
Ok, thank you very much for this hint. I increased the value to 3600 and will report the result.

**Comment by ArtificialOwl** [comment]
I should also be able to slightly lighter some SQL requests fixing this, thanks for your report.

About the fact that there is no OwnerId, this should only occurs in the case of a file shared to a group_folder. If this is not the case, this is a bug.

I will release a new version this week end of fulltextsearch and files_fulltextsearch and we'll check if you have the same issues

**Comment by ArtificialOwl** [comment]
Please close if 1.0.2 fix this issue.

**Comment by herrep** [comment]
Regarding no OwnerId: The files to be indexed are located on an external storage which is addressed in Nextcloud via SMB/CIFS. Is this the reason that no OwnerId is available? Needs anything to be done to obtain an OwnerId for files located on an external storage?

**Comment by herrep** [comment]
After setting TICK_TTL to 3600 seconds, ./occ fulltextsearch:index is still ongoing and did not lead to a Force quit. Memory usage is at 650 MB and I can scroll around the results and errors. However, the indexing process is at Action "compareWithCurrentIndex" and progress remains at 0/274964 without any change. This state is now unchanged for hours.

The indexing shows around 100,000 errors which are about 20% of the indexed documents. The error is always the same:
Exception: Elasticsearch\Common\Exceptions\BadRequest400Exception
Message: pipeline with id [attachment] does not exist

What should I do with the index making now progress at all?

**Comment by ArtificialOwl** [comment]
can you try to upgarde to 1.0.2 before starting the index again ?

**Comment by herrep** [comment]
I upgraded to 1.0.2 and started the index again. Then, after the while, everything seems to be processed, as there is no progress any more. See screen shot:
<img width="499" alt="bildschirmfoto 2018-10-19 um 23 31 57" src="https://user-images.githubusercontent.com/41866019/47244862-62a28100-d3f7-11e8-929f-62c7dcb1284c.png">

It appears that all indexed documents have contentSize = 0. Does this mean that my index only covers the file names, but no content at all? What needs to be done so that the content of my files is indexed?

**Comment by blizzz** [comment]
@daita @herrep  Is there anything left to do here?

**Comment by herrep** [comment]
I think this issue can be closed. At present, I still have issues with "Force Quit" but this is quite likely because of lack of physical resources. I am about to move Nextcloud and Elasticsearch to another server with more resources. Then I may report again if issues remain.

**Comment by TorstenE** [comment]
Hallo Nach dem Start von elasticsearch erhalte ich nachfolgende Meldung.
Hat jemand eine idee, woran es liegen kann ?

Danke

Torsten.E

`
sudo -u www-data php /var/www/html/nextcloud/occ fulltextsearch:index


In Connection.php line 615:

  {"error":{"root_cause":[{"type":"mapper_parsing_exception","reason":"Root m
  apping definition has unsupported parameters:  [standard : {dynamic=true, p
  roperties={owner={type=keyword}, groups={type=keyword}, circles={type=keywo
  rd}, metatags={type=keyword}, source={type=keyword}, title={copy_to=combine
  d, analyzer=keyword, term_vector=yes, type=text}, content={copy_to=combined
  , analyzer=analyzer, term_vector=yes, type=text}, users={type=keyword}, tag
  s={type=keyword}, provider={type=keyword}, subtags={type=keyword}, links={t
  ype=keyword}, combined={analyzer=analyzer, term_vector=yes, type=text}, has
  h={type=keyword}}}]"}],"type":"mapper_parsing_exception","reason":"Failed t
  o parse mapping [_doc]: Root mapping definition has unsupported parameters:
    [standard : {dynamic=true, properties={owner={type=keyword}, groups={type
  =keyword}, circles={type=keyword}, metatags={type=keyword}, source={type=ke
  yword}, title={copy_to=combined, analyzer=keyword, term_vector=yes, type=te
  xt}, content={copy_to=combined, analyzer=analyzer, term_vector=yes, type=te
  xt}, users={type=keyword}, tags={type=keyword}, provider={type=keyword}, su
  btags={type=keyword}, links={type=keyword}, combined={analyzer=analyzer, te
  rm_vector=yes, type=text}, hash={type=keyword}}}]","caused_by":{"type":"map
  per_parsing_exception","reason":"Root mapping definition has unsupported pa
  rameters:  [standard : {dynamic=true, properties={owner={type=keyword}, gro
  ups={type=keyword}, circles={type=keyword}, metatags={type=keyword}, source
  ={type=keyword}, title={copy_to=combined, analyzer=keyword, term_vector=yes
  , type=text}, content={copy_to=combined, analyzer=analyzer, term_vector=yes
  , type=text}, users={type=keyword}, tags={type=keyword}, provider={type=key
  word}, subtags={type=keyword}, links={type=keyword}, combined={analyzer=ana
  lyzer, term_vector=yes, type=text}, hash={type=keyword}}}]"}},"status":400}


fulltextsearch:index [--output [OUTPUT]] [-r|--no-readline] [--] [<options>]
`

**Comment by Omaha2002** [comment]
sudo -u www-data php /var/www/html/nextcloud/occ fulltextsearch:index

In Connection.php line 615:

{"error":{"root_cause":[{"type":"mapper_parsing_exception",.....

Ditto get the same error, freshly installed Elasticsearch 7 on openjdk 1.8.0_191

**Comment by af9** [comment]
@TorstenE exactly same result I get too with my NC 15.0.7 and elasticsearch 7.0 (see below).
The problem seems to be in the mapping for [_doc] ...
I think we should open a new issue, since our problem seems to be different to @herrep's one, which is already closed btw.

`{
	"error": {
		"root_cause": [{
				"type": "mapper_parsing_exception",
				"reason": "Root mapping definition has unsupported parameters:  [standard : {dynamic=true, properties={owner={type=keyword}, groups={type=keyword}, 
				circles={type=keyword}, metatags={type=keyword}, source={type=keyword}, title={copy_to=combined, analyzer=keyword, term_vector=yes, type=text}, 
				content={copy_to=combined, analyzer=analyzer, term_vector=yes, type=text}, users={type=keyword}, tags={type=keyword}, provider={type=keyword}, 
				subtags={type=keyword}, links={type=keyword}, combined={analyzer=analyzer, term_vector=yes, type=text}, hash={type=keyword}}}]"
			}
		],
		"type": "mapper_parsing_exception",
		"reason": "Failed to parse mapping [_doc]: Root mapping definition has unsupported parameters:  [standard : {dynamic=true, properties={owner={type=keyword}, 
		groups={type=keyword}, circles={type=keyword}, metatags={type=keyword}, source={type=keyword}, title={copy_to=combined, analyzer=keyword, term_vector=yes, type=text}, 
		content={copy_to=combined, analyzer=analyzer, term_vector=yes, type=text}, users={type=keyword}, tags={type=keyword}, provider={type=keyword}, subtags={type=keyword}, 
		links={type=keyword}, combined={analyzer=analyzer, term_vector=yes, type=text}, hash={type=keyword}}}]",
		"caused_by": {
			"type": "mapper_parsing_exception",
			"reason": "Root mapping definition has unsupported parameters:  [standard : {dynamic=true, properties={owner={type=keyword}, groups={type=keyword}, 
			circles={type=keyword}, metatags={type=keyword}, source={type=keyword}, title={copy_to=combined, analyzer=keyword, term_vector=yes, type=text}, 
			content={copy_to=combined, analyzer=analyzer, term_vector=yes, type=text}, users={type=keyword}, tags={type=keyword}, provider={type=keyword},
			subtags={type=keyword}, links={type=keyword}, combined={analyzer=analyzer, term_vector=yes, type=text}, hash={type=keyword}}}]"
		}
	},
	"status": 400
}`
